astropy/io/ascii/core.py

```diff
@@ -1036,13 +1037,14 @@ def _validate_and_copy(col, converters):
 
     def _convert_vals(self, cols):
         for col in cols:
-            # If a specific dtype was specified for a column, then use that
-            # to set the defaults, otherwise use the generic defaults.
-            default_converters = ([convert_numpy(col.dtype)] if col.dtype
-                                  else self.default_converters)
-
-            # If the user supplied a specific convert then that takes precedence over defaults
-            converters = self.converters.get(col.name, default_converters)
+            if col.dtype is not None:
```

Comment:
I don't think this is right - a user-given converter should always have preference over something inferred from the dtype - I think this belongs in the `else` clause of the loop over user converters.

Reply:
This is a line from @taldcroft that I indeed struggled to understand - so I left it in. I have not been able to find a situation where `col.dtype` is different from `None` at this point. I have not looked very deep, but it's hard to track where `dtype` is set. ("grep -r dtype *" is not very specific.) I'll try to look more - specifically since I now realize that we can roll #11278 in here, too.

Reply:
Indeed, that would be nice to roll in!  I do think the check for `col.dtype` is not right as is, but I'd better let @taldcroft pipe in...

Reply:
I made it raise an exception if I ever get to that point and in > 700 tests that never happens. So, no need to worry about this code path right now, but that is where I'll wire in the solution for #11278.

Reply:
As an aside: Since col.dtype does not seem to be used, I'm not sure we need it at all. In fact, I'm not sure we need the entire `core.Column` class as a class. A dict would do for that purpose, but that's a different matter.

Reply:
I agree the logic in my original suggestion was wrong, sorry about that, I just did it too quickly.

About `col.dtype`, this is used by `ecsv.py`, but since 4.3 the ECSV parser overrides the `_convert_vals` so that is why there is no longer any use of it in the base TableOutputter. However, since `Column.dtype` had previously been in the code for a while, it is possible that someone has their own parser that uses it. So I would suggest to maintain this for now.

Reply:
It's certainly possible. But if none of the people on this thread did that, it's extremely unlikely that anybody else does; I mean even people who are highly experienced in astropy (e.g. @pllim, @bsipocz, @hamogu, @mhvk) don't easily see through all the intricacies of `io.ascii`. So, I very much doubt that anyone else read the code for this mostly undocumented (and most unused) internal field in the `Column` class that nobody knows about to build something on top of it. How do I even deprecate an attribute? I'll see if I can up with that way that's not too convoluted to keep it in, but also keep everything else in the right order.

Reply:
I would still advocate for a new `dtype` kwarg for io, and slowly deprecate `converters`. Since I commented on the original issue, I run into it a few more times with users where they did a pandas read in order to be able to specify the dtype and straight away converting it into a Table, rather than using astropy all the way. Having `dtype` and a bit of documentation would eliminate at least a fraction of these unnecessary workarounds.

Reply:
There is a new dtype keyword: https://github.com/astropy/astropy/pull/11892/files#diff-cc98bdb28c2e64028b41ea31699d985af24fe744e6d5b1dde61359807a173447R51
This discussion is about *also* keeping the dtype attribute because someone may have used it.

Reply:
Yeah, I have no opinion on the attribute. What's the long term plan with the converters kwarg though? That would need a depreciation, wouldn't it?

Reply:
I think the converters attribute is there to stay. The addition of `dtype`in this form (list with access by position, not my name), is for simplicity and to match the `Table(... dtype=...)` interface, but the converters allow more flexibility (for once, they allow any callable, just just numpy dtypes). I don't think we should deprecate or remove converters, but we can reduce the amount of advertising. Everything that has been done with converters could have been done by deriving an outputter class and overriding `_convert_vals`, yet neither of users in #4934 and #11278 made a new class - both issues are about keyword arguments. So, I think we want to keep the keyword argument `converters` for full flexibility.

Reply:
I'm not talking about the converters attribute, but about the kwarg, which seemed to have some momentum to finally deprecate as being difficult to discover and doesn't match any API at all either in astropy or in other highly used packages. I'm away from keyboard so it's difficult to search the issues, but there were a thread a couple of months ago on the old 4xyz issue

Reply:
I'm sorry, but I don't understand, probably because I don't know all the context about recent developments. Maybe @taldcroft can help with that.

I'm saying that I would prefer the converters kwarg to stay - in fact this whole PR started by adding `fnmatch` capabilities to the converters kwarg. That's already the second iteration of an "improve the converters API" PR, the previous one worked with defaultdics for converters, but was turned down by @taldcroft and @mhvk.